src/ips/algorithm.rs

```diff
@@ -77,30 +80,64 @@ impl Ips {
 
     /// Generate peer list - main function with The Algorithm
     pub async fn generate(&mut self, state: &CrunchyState) -> Vec<Peer> {
+        // Set up logging
+        let output = match self.config.log_path {
+            Some(ref path) => File::create(path).map(|f| Box::new(f) as Box<dyn Write>),
+            None => Ok(Box::new(io::stdout()) as Box<dyn Write>),
+        };
+
+        let mut output = match output {
+            Ok(output) => output,
+            Err(e) => {
+                println!("Failed to open log file: {}", e);
+                Box::new(io::stdout()) as Box<dyn Write>
+            }
+        };
+
         // Sanity check that each node is really connected to its peers and the peers also
         // have the node in their connections.
+        writeln!(output, "IPS algorithm started... ").unwrap();
+        let start_time = std::time::Instant::now();
+
+        writeln!(output, "Checking for nodes connected to themselves...").unwrap();
         for (idx, node) in state.nodes.iter().enumerate() {
             if node.connections.contains(&idx) {
-                println!("{} is connected to itself.", node.addr);
+                writeln!(output, "{} is connected to itself.", node.addr).unwrap();
             }
 
             for peer in &node.connections {
                 if !state.nodes[*peer].connections.contains(&idx) {
-                    println!(
+                    writeln!(
+                        output,
                         "{} is not connected to {} but {} have a connection to it",
                         node.addr, state.nodes[*peer].addr, node.addr
-                    );
+                    )
+                    .unwrap();
                 }
             }
         }
 
+        writeln!(
+            output,
+            "Generating initial network state and its statistics... "
+        )
+        .unwrap();
+
         // This is the working set of factors.
         let mut working_state = self.generate_state(&state.nodes, false);
         let mut final_state = working_state.clone();
 
         let initial_statistics = generate_statistics(&working_state);
-        println!("Statistics for the initial network:");
-        print_statistics(&initial_statistics);
+
+        writeln!(output, "Statistics for the initial network:").unwrap();
+        print_statistics(&initial_statistics, &mut output);
```

Comment:
to reflect the above `std::fs` APIs, the `fs`/`ouput` variable always is a first param
```suggestion
        print_statistics(&mut output, &initial_statistics);
```

Reply:
Changed.